src/tail/session.rs

```diff
@@ -66,125 +65,79 @@ impl Session {
-                tx.send(()).unwrap();
-                failure::bail!(http::format_error(e, Some(&tail_help)))
-            }
+        } else {
+            failure::bail!("Could not extract tunnel url from cloudflared");
         }
     }
 }
 
-async fn get_tunnel_url(metrics_port: u16) -> Result<String, failure::Error> {
-    // regex for extracting url from cloudflared metrics port.
+async fn get_tunnel_url(metrics_port: u16) -> Result<String, ()> {
+    let metrics_url = format!("http://localhost:{}/metrics", metrics_port);
     let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
-
-    struct RetryDelay {
-        delay: Sleep,
-        attempt: u64,
-        max_attempts: u64,
-    }
-
-    // This retry loop retries retrieving the cloudflared endpoint url from the cloudflared /metrics
-    // until it gets the URL or has tried retrieving the URL over 5 times.
-    impl RetryDelay {
-        fn new(max_attempts: u64) -> RetryDelay {
-            RetryDelay {
-                delay: sleep(Duration::from_millis(0)),
-                attempt: 0,
-                max_attempts,
-            }
-        }
-
-        // our retry delay is an [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff),
-        // which simply waits twice as long between each attempt to avoid hammering the LogServer.
-        fn reset(self) -> RetryDelay {
-            let attempt = self.attempt + 1;
-            let delay = sleep(Duration::from_millis(attempt * attempt * 1000));
-            let max_attempts = self.max_attempts;
-
-            RetryDelay {
-                attempt,
-                delay,
-                max_attempts,
-            }
-        }
-
-        fn expired(&self) -> bool {
-            self.attempt > self.max_attempts
-        }
-
-        fn is_elapsed(&self) -> bool {
-            self.delay.is_elapsed()
-        }
-    }
-
-    let mut delay = RetryDelay::new(5);
-
-    while !delay.expired() {
-        if delay.is_elapsed() {
-            let metrics_url = format!("http://localhost:{}/metrics", metrics_port);
-            if let Ok(resp) = reqwest::get(&metrics_url).await {
-                let body = resp.text().await?;
-
-                if let Some(capture) = url_regex.captures(&body) {
-                    if let Some(url) = capture.get(1) {
-                        return Ok(url.as_str().to_string());
-                    }
-                }
-            }
-
-            delay = delay.reset();
+    match reqwest::get(metrics_url).await {
+        Ok(resp) => match resp.text().await {
+            Ok(body) => match url_regex.captures(&body) {
+                Some(capture) => match capture.get(1) {
+                    Some(url) => Ok(url.as_str().to_string()),
+                    None => Err(()),
+                },
+                None => return Err(()),
+            },
+            Err(_) => return Err(()),
+        },
+        Err(_) => {
+            return Err(());
```

Comment:
This can be cleaned up a lot with `?` operators. e.g.:

```rust
let body = reqwest::get(metrics_url).await?.text().await?;
match url_regex.captures(&body).and_then(|capture| capture.get(1)) {
    Some(capture) => Ok(url.as_str().to_string()),
    None => Err(failure::err_msg!("Failed to extract capture group!")),
}
```

For more info: 
- [`and_then`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.and_then)
- [`?`](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html)

Reply:
Nice, I didn't know about `and_then`